File: packages/iroh-http-shared/src/session.ts
/**
 * `IrohSession` — a WebTransport-compatible session to a single remote peer.
 *
 * Created via `node.connect(peer)`.  Wraps a QUIC connection and exposes
 * bidirectional streams, unidirectional streams, and datagrams through
 * the standard WebTransport interface.
 */

import type { FfiDuplexStream, IrohAdapter } from "./IrohAdapter.js";
import type { PublicKey } from "./keys.js";
import { makeReadable } from "./streams.js";

/** WebTransport-compatible bidirectional stream. */
export interface WebTransportBidirectionalStream {
  readonly readable: ReadableStream<Uint8Array>;
  readonly writable: WritableStream<Uint8Array>;
}

/** WebTransport close info. */
export interface WebTransportCloseInfo {
  closeCode: number;
  reason: string;
}

/** WebTransport datagram duplex stream. */
export interface WebTransportDatagramDuplexStream {
  readonly readable: ReadableStream<Uint8Array>;
  readonly writable: WritableStream<Uint8Array>;
  readonly maxDatagramSize: number | null;
  incomingHighWaterMark: number;
  outgoingHighWaterMark: number;
}

/** Raw session FFI functions provided by each platform adapter. */
export interface RawSessionFns {
  /** Establish a session to a remote peer. Returns an opaque session handle. */
  connect(
    endpointHandle: number,
    nodeId: string,
    directAddrs: string[] | null,
  ): Promise<bigint>;
  /**
   * Accept an incoming session from a remote peer.
   *
   * Blocks until a peer opens a raw QUIC connection.  Returns the session
   * handle and the remote peer's node ID string, or `null` when the endpoint
   * is shutting down.
   */
  sessionAccept?(
    endpointHandle: number,
  ): Promise<{ sessionHandle: bigint; nodeId: string } | null>;
  /** Open a new bidi stream on an existing session. */
  createBidiStream(sessionHandle: bigint): Promise<FfiDuplexStream>;
  /** Accept the next incoming bidi stream. Returns `null` when the session closes. */
  nextBidiStream(sessionHandle: bigint): Promise<FfiDuplexStream | null>;
  /** Open a new unidirectional (send-only) stream. Returns a write handle. */
  createUniStream(sessionHandle: bigint): Promise<bigint>;
  /** Accept the next incoming unidirectional (receive-only) stream. Returns a read handle, or `null` when closed. */
  nextUniStream(sessionHandle: bigint): Promise<bigint | null>;
  /** Send a datagram. */
  sendDatagram(sessionHandle: bigint, data: Uint8Array): Promise<void>;
  /** Receive the next datagram. Returns `null` when the session closes. */
  recvDatagram(sessionHandle: bigint): Promise<Uint8Array | null>;
  /** Get the maximum datagram payload size. Returns `null` if datagrams are unsupported. */
  maxDatagramSize(sessionHandle: bigint): Promise<number | null>;
  /** Wait for the session to close. Returns close info. */
  closed(sessionHandle: bigint): Promise<WebTransportCloseInfo>;
  /** Close the session with an optional close code and reason. */
  close(
    sessionHandle: bigint,
    closeCode?: number,
    reason?: string,
  ): Promise<void>;
}

/**
 * A session to a single remote peer.
 *
 * Implements the WebTransport interface.
 */
export interface IrohSession {
  /** The remote peer's public key. */
  readonly remoteId: PublicKey;

  /** Resolves when the QUIC handshake completes. */
  readonly ready: Promise<undefined>;

  /**
   * Open a new bidirectional stream.
   */
  createBidirectionalStream(): Promise<WebTransportBidirectionalStream>;

  /**
   * Open a new unidirectional (send-only) stream.
   */
  createUnidirectionalStream(): Promise<WritableStream<Uint8Array>>;

  /**
   * Incoming bidirectional streams opened by the remote peer.
   */
  readonly incomingBidirectionalStreams: ReadableStream<
    WebTransportBidirectionalStream
  >;

  /**
   * Incoming unidirectional streams from the remote peer.
   */
  readonly incomingUnidirectionalStreams: ReadableStream<
    ReadableStream<Uint8Array>
  >;

  /** Datagram duplex stream for unreliable message passing. */
  readonly datagrams: WebTransportDatagramDuplexStream;

  /** Resolves when the session is fully closed, with close code and reason. */
  readonly closed: Promise<WebTransportCloseInfo>;

  /** Close the session with an optional close code and reason. */
  close(closeInfo?: WebTransportCloseInfo): void;

  /** TC39 explicit resource management. */
  [Symbol.asyncDispose](): Promise<void>;
}

/**
 * Build an `IrohSession` from raw platform handles.
 */
export function buildSession(
  adapter: IrohAdapter,
  sessionHandle: bigint,
  remoteId: PublicKey,
  rawSession: RawSessionFns,
): IrohSession {
  // The session_closed promise from the native side.
  const closedPromise = rawSession.closed(sessionHandle);

  function wrapDuplex(ffi: FfiDuplexStream): WebTransportBidirectionalStream {
    const readable = makeReadable(adapter, ffi.readHandle);
    const writable = new WritableStream<Uint8Array>({
      async write(chunk) {
        await adapter.sendChunk(ffi.writeHandle, chunk);
      },
      async close() {
        await adapter.finishBody(ffi.writeHandle);
      },
      async abort() {
        await adapter.finishBody(ffi.writeHandle);
      },
    });
    return { readable, writable };
  }

  // Lazy incoming bidi streams — only one ReadableStream instance.
  let _incomingBidiStreams:
    | ReadableStream<WebTransportBidirectionalStream>
    | null = null;

  // Lazy incoming uni streams.
  let _incomingUniStreams: ReadableStream<ReadableStream<Uint8Array>> | null =
    null;

  // Lazy datagram duplex stream.
  let _datagrams: WebTransportDatagramDuplexStream | null = null;

  const session: IrohSession = {
    remoteId,

    ready: Promise.resolve(undefined),

    async createBidirectionalStream(): Promise<
      WebTransportBidirectionalStream
    > {
      const ffi = await rawSession.createBidiStream(sessionHandle);
      return wrapDuplex(ffi);
    },

    async createUnidirectionalStream(): Promise<WritableStream<Uint8Array>> {
      const writeHandle = await rawSession.createUniStream(sessionHandle);
      return new WritableStream<Uint8Array>({
        async write(chunk) {
          await adapter.sendChunk(writeHandle, chunk);
        },
        async close() {
          await adapter.finishBody(writeHandle);
        },
        async abort() {
          await adapter.finishBody(writeHandle);
        },
      });
    },

    get incomingBidirectionalStreams(): ReadableStream<
      WebTransportBidirectionalStream
    > {
      if (!_incomingBidiStreams) {
        _incomingBidiStreams = new ReadableStream<
          WebTransportBidirectionalStream
        >({
          async pull(controller) {
            const ffi = await rawSession.nextBidiStream(sessionHandle);
            if (ffi === null) {
              controller.close();
            } else {
              controller.enqueue(wrapDuplex(ffi));
            }
          },
        });
      }
      return _incomingBidiStreams;
    },

    get incomingUnidirectionalStreams(): ReadableStream<
      ReadableStream<Uint8Array>
    > {
      if (!_incomingUniStreams) {
        _incomingUniStreams = new ReadableStream<ReadableStream<Uint8Array>>({
          async pull(controller) {
            const readHandle = await rawSession.nextUniStream(sessionHandle);
            if (readHandle === null) {
              controller.close();
            } else {
              controller.enqueue(makeReadable(adapter, readHandle));
            }
          },
        });
      }
      return _incomingUniStreams;
    },

    get datagrams(): WebTransportDatagramDuplexStream {
      if (!_datagrams) {
        const readable = new ReadableStream<Uint8Array>({
          async pull(controller) {
            const data = await rawSession.recvDatagram(sessionHandle);
            if (data === null) {
              controller.close();
            } else {
              controller.enqueue(data);
            }
          },
        });

        const writable = new WritableStream<Uint8Array>({
          async write(chunk) {
            await rawSession.sendDatagram(sessionHandle, chunk);
          },
        });

        let maxSize: number | null = null;
        // Eagerly fetch max datagram size (fire-and-forget).
        void rawSession.maxDatagramSize(sessionHandle).then((s) => {
          maxSize = s;
        });

        _datagrams = {
          readable,
          writable,
          get maxDatagramSize() {
            return maxSize;
          },
          incomingHighWaterMark: 1,
          outgoingHighWaterMark: 1,
        };
      }
      return _datagrams;
    },

    closed: closedPromise,

    close(closeInfo?: WebTransportCloseInfo): void {
      void rawSession
        .close(
          sessionHandle,
          closeInfo?.closeCode ?? 0,
          closeInfo?.reason || undefined,
        )
        .catch(() => {});
    },

    [Symbol.asyncDispose]() {
      session.close();
      return closedPromise.then(() => {});
    },
  };

  return session;
}

File: crates/iroh-http-core/src/session.rs
//! Session — a QUIC connection to a single remote peer.
//!
//! `session_connect` establishes (or reuses a pooled) connection and returns
//! an opaque handle.  Bidirectional streams, unidirectional streams, and
//! datagrams are all accessible through the session handle.

use iroh::endpoint::Connection;
use serde::Serialize;

use crate::{
    parse_node_addr,
    stream::{pump_body_to_quic_send, pump_quic_recv_to_body, HandleStore, SessionEntry},
    CoreError, FfiDuplexStream, IrohEndpoint, ALPN_DUPLEX,
};

/// Returns `true` if the connection error means "connection ended" rather
/// than a protocol-level bug.  Used to return `None` instead of `Err`.
fn is_connection_closed(err: &iroh::endpoint::ConnectionError) -> bool {
    use iroh::endpoint::ConnectionError::*;
    matches!(
        err,
        ApplicationClosed(_) | ConnectionClosed(_) | Reset | TimedOut | LocallyClosed
    )
}

/// Close information returned when a session ends.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CloseInfo {
    pub close_code: u64,
    pub reason: String,
}

// ── Session registry ──────────────────────────────────────────────────────────

fn get_conn(endpoint: &IrohEndpoint, handle: u64) -> Result<Connection, CoreError> {
    endpoint
        .handles()
        .lookup_session(handle)
        .map(|s| s.conn.clone())
        .ok_or_else(|| CoreError::invalid_handle(handle))
}

/// Return the remote peer's public key for a session.
pub fn session_remote_id(
    endpoint: &IrohEndpoint,
    handle: u64,
) -> Result<iroh::PublicKey, CoreError> {
    get_conn(endpoint, handle).map(|c| c.remote_id())
}

// ── Public API ────────────────────────────────────────────────────────────────

/// Establish a session (QUIC connection) to a remote peer.
///
/// Each call creates a **dedicated** QUIC connection — sessions are not pooled.
/// This ensures that closing one session handle cannot affect other sessions
/// to the same peer.  (Fetch operations continue to use the shared pool for
/// efficiency; sessions opt out because `session_close` closes the underlying
/// connection.)
///
/// Returns an opaque session handle.
pub async fn session_connect(
    endpoint: &IrohEndpoint,
    remote_node_id: &str,
    direct_addrs: Option<&[std::net::SocketAddr]>,
) -> Result<u64, CoreError> {
    let parsed = parse_node_addr(remote_node_id)?;
    let node_id = parsed.node_id;
    let mut addr = iroh::EndpointAddr::new(node_id);
    for a in &parsed.direct_addrs {
        addr = addr.with_ip_addr(*a);
    }
    if let Some(addrs) = direct_addrs {
        for a in addrs {
            addr = addr.with_ip_addr(*a);
        }
    }

    let conn = endpoint
        .raw()
        .connect(addr, ALPN_DUPLEX)
        .await
        .map_err(|e| CoreError::connection_failed(format!("connect session: {e}")))?;

    let handle = endpoint.handles().insert_session(SessionEntry { conn })?;

    Ok(handle)
}

/// Open a new bidirectional stream on an existing session.
///
/// Returns `FfiDuplexStream` with `read_handle` / `write_handle` backed by
/// body channels — the same interface used by fetch and raw_connect.
pub async fn session_create_bidi_stream(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<FfiDuplexStream, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;

    let (send, recv) = conn
        .open_bi()
        .await
        .map_err(|e| CoreError::connection_failed(format!("open_bi: {e}")))?;

    wrap_bidi_stream(endpoint.handles(), send, recv)
}

/// Accept the next incoming bidirectional stream from the remote side.
///
/// Blocks until the remote opens a stream, or returns `None` when the
/// connection is closed.
pub async fn session_next_bidi_stream(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<Option<FfiDuplexStream>, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;

    match conn.accept_bi().await {
        Ok((send, recv)) => Ok(Some(wrap_bidi_stream(endpoint.handles(), send, recv)?)),
        Err(e) if is_connection_closed(&e) => Ok(None),
        Err(e) => Err(CoreError::connection_failed(format!("accept_bi: {e}"))),
    }
}

/// Accept an incoming session (QUIC connection) from a remote peer.
///
/// Blocks until a peer connects.  Returns an opaque session handle, or
/// `None` if the endpoint is shutting down.
pub async fn session_accept(endpoint: &IrohEndpoint) -> Result<Option<u64>, CoreError> {
    let incoming = match endpoint.raw().accept().await {
        Some(inc) => inc,
        None => return Ok(None),
    };

    let conn = incoming
        .await
        .map_err(|e| CoreError::connection_failed(format!("accept session: {e}")))?;

    let handle = endpoint.handles().insert_session(SessionEntry { conn })?;

    Ok(Some(handle))
}

/// Close a session and remove it from the registry.
///
/// `close_code` is an application-level error code (maps to QUIC VarInt).
/// `reason` is a human-readable string sent to the peer.
pub fn session_close(
    endpoint: &IrohEndpoint,
    session_handle: u64,
    close_code: u64,
    reason: &str,
) -> Result<(), CoreError> {
    let entry = endpoint
        .handles()
        .remove_session(session_handle)
        .ok_or_else(|| CoreError::invalid_handle(session_handle))?;
    let code = iroh::endpoint::VarInt::from_u64(close_code).map_err(|_| {
        CoreError::invalid_input(format!(
            "close_code {close_code} exceeds QUIC VarInt max (2^62 - 1)"
        ))
    })?;
    entry.conn.close(code, reason.as_bytes());
    Ok(())
}

/// Wait for the QUIC handshake to complete on a session.
///
/// Resolves immediately if the handshake has already completed.
pub async fn session_ready(endpoint: &IrohEndpoint, session_handle: u64) -> Result<(), CoreError> {
    // Validate handle exists — keeps error behavior consistent with other session APIs.
    let _conn = get_conn(endpoint, session_handle)?;
    // iroh connections are fully established by the time session_connect returns,
    // so ready always resolves immediately. Kept for WebTransport API compatibility.
    Ok(())
}

/// Wait for the session to close and return the close information.
///
/// Blocks until the connection is closed by either side.  Removes the
/// session from the registry so resources are freed.
pub async fn session_closed(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<CloseInfo, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;
    let err = conn.closed().await;
    // Connection is dead — clean up the registry entry.
    endpoint.handles().remove_session(session_handle);
    let (close_code, reason) = parse_connection_error(&err);
    Ok(CloseInfo { close_code, reason })
}

// ── Unidirectional streams ────────────────────────────────────────────────────

/// Open a new unidirectional (send-only) stream on an existing session.
///
/// Returns a write handle backed by a body channel.
pub async fn session_create_uni_stream(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<u64, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;
    let send = conn
        .open_uni()
        .await
        .map_err(|e| CoreError::connection_failed(format!("open_uni: {e}")))?;

    let handles = endpoint.handles();
    let (send_writer, send_reader) = handles.make_body_channel();
    let write_handle = handles.insert_writer(send_writer)?;
    tokio::spawn(pump_body_to_quic_send(send_reader, send));

    Ok(write_handle)
}

/// Accept the next incoming unidirectional (receive-only) stream.
///
/// Returns a read handle, or `None` when the connection is closed.
pub async fn session_next_uni_stream(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<Option<u64>, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;

    match conn.accept_uni().await {
        Ok(recv) => {
            let handles = endpoint.handles();
            let (recv_writer, recv_reader) = handles.make_body_channel();
            let read_handle = handles.insert_reader(recv_reader)?;
            tokio::spawn(pump_quic_recv_to_body(recv, recv_writer));
            Ok(Some(read_handle))
        }
        Err(e) if is_connection_closed(&e) => Ok(None),
        Err(e) => Err(CoreError::connection_failed(format!("accept_uni: {e}"))),
    }
}

// ── Datagrams ─────────────────────────────────────────────────────────────────

/// Send a datagram on the session.
///
/// Fails if `data.len()` exceeds `session_max_datagram_size`.
pub fn session_send_datagram(
    endpoint: &IrohEndpoint,
    session_handle: u64,
    data: &[u8],
) -> Result<(), CoreError> {
    let conn = get_conn(endpoint, session_handle)?;
    conn.send_datagram(bytes::Bytes::copy_from_slice(data))
        .map_err(|e| CoreError::internal(format!("send_datagram: {e}")))
}

/// Receive the next datagram from the session.
///
/// Blocks until a datagram arrives, or returns `None` when the connection closes.
pub async fn session_recv_datagram(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<Option<Vec<u8>>, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;
    match conn.read_datagram().await {
        Ok(data) => Ok(Some(data.to_vec())),
        Err(e) if is_connection_closed(&e) => Ok(None),
        Err(e) => Err(CoreError::connection_failed(format!("recv_datagram: {e}"))),
    }
}

/// Return the current maximum datagram payload size for this session.
///
/// Returns `None` if datagrams are not supported on the current path.
pub fn session_max_datagram_size(
    endpoint: &IrohEndpoint,
    session_handle: u64,
) -> Result<Option<usize>, CoreError> {
    let conn = get_conn(endpoint, session_handle)?;
    Ok(conn.max_datagram_size())
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/// Wrap raw QUIC send/recv streams into body-channel–backed `FfiDuplexStream`.
fn wrap_bidi_stream(
    handles: &HandleStore,
    send: iroh::endpoint::SendStream,
    recv: iroh::endpoint::RecvStream,
) -> Result<FfiDuplexStream, CoreError> {
    let mut guard = handles.insert_guard();

    // Receive side: pump from QUIC recv → BodyWriter → BodyReader (JS reads via nextChunk).
    let (recv_writer, recv_reader) = handles.make_body_channel();
    let read_handle = guard.insert_reader(recv_reader)?;
    tokio::spawn(pump_quic_recv_to_body(recv, recv_writer));

    // Send side: pump from BodyReader (JS writes via sendChunk) → QUIC send.
    let (send_writer, send_reader) = handles.make_body_channel();
    let write_handle = guard.insert_writer(send_writer)?;
    tokio::spawn(pump_body_to_quic_send(send_reader, send));

    guard.commit();
    Ok(FfiDuplexStream {
        read_handle,
        write_handle,
    })
}

/// Extract close code and reason from a QUIC `ConnectionError`.
fn parse_connection_error(err: &iroh::endpoint::ConnectionError) -> (u64, String) {
    match err {
        iroh::endpoint::ConnectionError::ApplicationClosed(info) => {
            let code: u64 = info.error_code.into();
            let reason = String::from_utf8_lossy(&info.reason).into_owned();
            (code, reason)
        }
        other => (0, other.to_string()),
    }
}
